feat: add early stopping/pruner for behavior tests cost optimization

ixchio opened the pull request:
## Summary
Closes #1417

Implements early stopping mechanism to detect failures early and terminate behavior tests before full trajectory completes, reducing LLM costs.

## Changes
- Add `early_stopper.py` with 6 pruner classes
- Integrate early stopping in `BaseIntegrationTest` callback  
- Add `get_early_stopper()` hook in `SoftwareAgentSDKBehaviorTest`
- Apply early stoppers to b01, b02, b05 behavior tests
- Add 17 unit tests for early stopper functionality

## Per discussion with @ryanhoangt
- Pattern-based pruning first (zero LLM cost)
- Stop on first failure signal
- Skip final LLM judge when early stopped
- Reusable pruner classes with base interface

## Test Results
All 17 unit tests passing

ryanhoangt reviewed (commented):
Thank you for the PR, I left a few comments below, and currently the linting workflow seems to fail. One thing I'm concerned about is we have too many concrete implementations of the pruners, which is a bit repetitive and hard to maintain. OTOH I still don't have a clear picture what the best design should look like. @xingyaoww do you have any thoughts about this?

---
This is a proposal from OH, which seems pretty good to me:

Here’s a concrete sketch for a simpler, composable early-stopping system you can drop 
into `tests/integration/early_stopper.py`. The key ideas:

* keep only two reusable pruners (`CommandPatternPruner` and `ActionLimitPruner`)
* provide a `compose` helper to combine them when needed
* let behavior tests configure early-stopping declaratively instead of subclassing

---

### Core primitives

```python
# tests/integration/early_stopper.py

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, Iterable

from openhands.sdk.event.base import Event
from openhands.sdk.event.llm_convertible.action import ActionEvent

from openhands.tools.file_editor.definition import FileEditorAction, FileEditorTool
from openhands.tools.terminal.definition import TerminalAction, TerminalTool


@dataclass
class EarlyStopResult:
    should_stop: bool
    reason: str | None = None


Checker = Callable[[Iterable[Event]], EarlyStopResult]


def compose(*checkers: Checker) -> Checker:
    """Return a checker that stops when any sub-checker stops."""
    def combined(events: Iterable[Event]) -> EarlyStopResult:
        for checker in checkers:
            result = checker(events)
            if result.should_stop:
                return result
        return EarlyStopResult(False)
    return combined
```

---

### Command-pattern pruner

```python
def CommandPatternPruner(
    tool: str,
    action_filter: Callable[[ActionEvent], bool],
    forbidden_patterns: list[str],
    *,
    reason_template: str,
) -> Checker:
    """Stop when an action's `command` contains a forbidden pattern."""

    def checker(events: Iterable[Event]) -> EarlyStopResult:
        for event in events:
            if isinstance(event, ActionEvent) and event.tool_name == tool:
                action = event.action
                if action is not None and action_filter(event):
                    command = getattr(action, "command", "")
                    for pattern in forbidden_patterns:
                        if pattern in command:
                            return EarlyStopResult(
                                True,
                                reason_template.format(pattern=pattern, 
command=command),
                            )
        return EarlyStopResult(False)

    return checker
```

Example usage (see below for b01/b02).

---

### Action-count limiter

```python
def ActionLimitPruner(
    tool: str,
    action_filter: Callable[[ActionEvent], bool],
    *,
    max_count: int,
    ignore: Callable[[ActionEvent], bool] | None = None,
    reason_template: str,
) -> Checker:
    """Stop when matching actions exceed `max_count`."""

    ignore = ignore or (lambda _: False)

    def checker(events: Iterable[Event]) -> EarlyStopResult:
        matching = 0
        for event in events:
            if isinstance(event, ActionEvent) and event.tool_name == tool:
                if event.action is None:
                    continue
                if ignore(event):
                    continue
                if action_filter(event):
                    matching += 1
                    if matching > max_count:
                        return EarlyStopResult(
                            True, reason_template.format(count=matching)
                        )
        return EarlyStopResult(False)

    return checker
``` 

#### b01 – “no premature implementation”

```python
def get_early_stopper(self):
    return CommandPatternPruner(
        tool=FileEditorTool.name,
        action_filter=lambda ev: isinstance(ev.action, FileEditorAction),
        forbidden_patterns=["create", "str_replace", "insert", "undo_edit"],
        reason_template=(
            "Agent attempted forbidden file edit: '{pattern}' in {command}"
        ),
    )
```

#### b02 – “no over-verification”

```python
def get_early_stopper(self):
    broad_tests = CommandPatternPruner(
        tool=TerminalTool.name,
        action_filter=lambda ev: isinstance(ev.action, TerminalAction)
        and ("pytest" in ev.action.command or "python -m unittest" in 
ev.action.command),
        forbidden_patterns=[
            "pytest tests/",
            "pytest .",
            "python -m pytest .",
            "pytest -x tests/",
        ],
        reason_template="Detected overly broad test command containing '{pattern}'",
    )

    test_spam = ActionLimitPruner(
        tool=TerminalTool.name,
        action_filter=lambda ev: isinstance(ev.action, TerminalAction)
        and ("pytest" in ev.action.command or "python -m unittest" in 
ev.action.command),
        ignore=lambda ev: "tests/tools/terminal" in ev.action.command,
        max_count=5,
        reason_template="Executed {count} test commands (limit 5)",
    )

    return compose(broad_tests, test_spam)
```

#### b05 – “no redundant files”

```python
def get_early_stopper(self):
    return ActionLimitPruner(
        tool=FileEditorTool.name,
        action_filter=lambda ev: isinstance(ev.action, FileEditorAction)
        and ev.action.command == "create",
        ignore=lambda ev: any(
            pattern in ev.action.path for pattern in (".openhands/", "__pycache__/")
        ),
        max_count=3,  # training script, optional README, optional test file
        reason_template="Created {count} files (limit 3)",
    )
```

ixchio commented:
hey @ryanhoangt - fixed both things you mentioned!

**trailing space pattern**: yea good catch, changed it to just `pytest tests/` without the trailing space. makes way more sense 👍

**mock tests**: rewrote them to use real ActionEvent/TerminalAction objects now. was kinda silly to test with mocks that bypass the actual implementation lol. all 17 tests pass with the real deal.

@enyst interesting point about #1467! that hooks system could def clean this up. lmk if you want me to wait for that or just go with this for now

enyst commented:
> @enyst interesting point about #1467! that hooks system could def clean this up. lmk if you want me to wait for that or just go with this for now

Oh, IMHO we could go with this just fine, thank you! 

I didn’t mean to imply waiting for hooks (though I think it’s ~mergeable), just that it will be great to put to good use. 😅 But it could be in a follow up refactoring.

I’ll defer to @ryanhoangt either way.

xingyaoww commented on tests/integration/early_stopper.py:
```diff
@@ -0,0 +1,357 @@
+            logger.warning("LLM judge pruner error: %s", e)
+            self._cached_result = EarlyStopResult(should_stop=False)
+
+        return self._cached_result
```
This is probably also an overkill too?

xingyaoww commented on tests/integration/tests/b05_do_not_create_redundant_files.py:
```diff
@@ -26,12 +28,77 @@
+                ),
+            )
+
+        return EarlyStopResult(should_stop=False)
```
I think this can also be a bit brittle

xingyaoww reviewed (commented):
I like the overall idea! But i do agree with @ryanhoangt that some of these early stopper could be a bit birttle.

@ixchio wdyt about removing the brittle ones, and we can get the base class and simpler earlystopper like `FileEditPruner` / `TerminalPruner` in?

xingyaoww commented on tests/integration/tests/b02_no_oververification.py:
```diff
@@ -33,6 +34,14 @@ class NoOververificationTest(SoftwareAgentSDKBehaviorTest):
+        by the final LLM judge rather than pattern-based pruning.
+        """
+        return None
+
```
Let's remove this, since it will default to None?
```suggestion
```
